From 9526bb0163f052a75950c3a63be1bd05e2814e8f Mon Sep 17 00:00:00 2001
From: Matthias Bolte <matthias@tinkerforge.com>
Date: Wed, 25 Oct 2023 14:06:24 +0200
Subject: [PATCH] wifi: Add setWriteNonBlocking option

Needed to stop ModbusTCP from blocking 10 seconds on a write in case
of connection loss.
---
 libraries/WiFi/src/WiFiClient.cpp | 2 +-
 libraries/WiFi/src/WiFiClient.h   | 2 ++
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/libraries/WiFi/src/WiFiClient.cpp b/libraries/WiFi/src/WiFiClient.cpp
index 85c34832..66fd85c5 100644
--- a/libraries/WiFi/src/WiFiClient.cpp
+++ b/libraries/WiFi/src/WiFiClient.cpp
@@ -405,7 +405,7 @@ size_t WiFiClient::write(const uint8_t *buf, size_t size)
         FD_ZERO(&set);        // empties the set
         FD_SET(socketFileDescriptor, &set); // adds FD to the set
         tv.tv_sec = 0;
-        tv.tv_usec = WIFI_CLIENT_SELECT_TIMEOUT_US;
+        tv.tv_usec = _writeNonBlocking ? 0 : WIFI_CLIENT_SELECT_TIMEOUT_US;
         retry--;
 
         if(select(socketFileDescriptor + 1, NULL, &set, NULL, &tv) < 0) {
diff --git a/libraries/WiFi/src/WiFiClient.h b/libraries/WiFi/src/WiFiClient.h
index b18f9324..fe7ac1e0 100644
--- a/libraries/WiFi/src/WiFiClient.h
+++ b/libraries/WiFi/src/WiFiClient.h
@@ -43,6 +43,7 @@ protected:
     std::shared_ptr<WiFiClientRxBuffer> _rxBuffer;
     bool _connected;
     int _timeout;
+    bool _writeNonBlocking;
 
 public:
     WiFiClient *next;
@@ -57,6 +58,7 @@ public:
     size_t write(const uint8_t *buf, size_t size);
     size_t write_P(PGM_P buf, size_t size);
     size_t write(Stream &stream);
+    void setWriteNonBlocking(bool nonBlocking) { _writeNonBlocking = nonBlocking; }
     int available();
     int read();
     int read(uint8_t *buf, size_t size);
-- 
2.34.1

